CMSgov/cms-carts-seds
# [Mdct 213] code auth flow & timeout modals

# Description
Switch over to code auth flow, which has the bonus result of handling refresh tokens for us and not timing out users arbitrarily.

Adds a modal to warn a user about idle timeouts. Additionally clears a users credentials when they come back to the app and they have been away longer than that idle time. Max idle time of 30 mins.

![image](https://user-images.githubusercontent.com/6362494/192625029-49117189-a302-4039-a8d5-3652a5298760.png)

![image](https://user-images.githubusercontent.com/6362494/192624944-f28e3859-dd8e-438a-b96e-beffd1b276c4.png)

**What changed in all those files?**
* For that nice modal, we're doing some logic to track idle timeout. We're grabbing on to history navigation, and api calls. This could be expanded in the future, but should be pretty reliable. Those API calls are now wrapped to just update our idle timeout.
* Config changes for implicit -> code flow. A little more secure, and also lets us lean on refresh tokens.
* New Timeout Modal silently sits on the page and renders if triggered by the react state.
* An auth lifecycle manager, that tracks our timeouts and signins.
  * Has two timers, one to show a modal on idle timeout, and if that is triggered, counts down to sign out the credentials.
  * Tracks if a user has been away from the app longer than the idle timer, so that the returning user needs to get a fresh session. This helps prevent a user from using the app at 2PM one day, and 1:45PM the next day and having their refresh token expire on them.
  * Also has a little bit of extra fluff to debounce the idle updates just so we're not blasting the redux store with user events.


## How to test
Log into the deployed branch. Wait 29 minutes, and see what happens.
https://d1fje2lk2rmsgb.cloudfront.net/


## Dependencies
Patience.

# Get it done

Now that the PR is open this is what happens next

## Code authors checklist


## Reviewers Checklist (Two different people)


## Assignee
